lib/client.js

```diff
@@ -553,6 +563,29 @@ Client.prototype.waitForSchemaAgreement = function (connection, callback) {
 };
 
 /**
+ * Waits for schema agreements and schedules schema metadata refresh.
+ * @param {Connection} connection
+ * @param event
+ * @param {Function} callback
+ * @ignore
+ * @internal
+ */
+Client.prototype.handleSchemaChange = function (connection, event, callback) {
+  var self = this;
+  this._waitForSchemaAgreement(connection, function (err) {
+    if (err) {
+      //we issue a warning but we continue with the normal flow
+      self.log('warning', 'There was an error while waiting for the schema agreement between nodes', err);
+    }
+    if (self.options.isMetadataEnabled) {
+      // schedule metadata refresh but do not wait for it to finish (no callback)
+      self.controlConnection.handleSchemaChange(event);
```

Comment:
Looks good, I think that solution solves 2 very well.  1 doesn't seem easily solveable (java driver has the same problem).  If the schema doesn't agree, the control node may not have the schema when it fetches it, however by the time the schema does agree, we'll get a schema_change even over the control connection, which will get us back up to date.  So I think the only weakness there is the possibility that the metadata in the driver is wrong immediately after the query completes, but I think that's ok given that the schema doesn't agree and when it does agree, it'll be up to date.